### PR TITLE
Disallow specified combinations of TLS-related connection string options

### PR DESCRIPTION
**Pull request by jyemin**
JAVA-3204

### CODE REVIEW

File: driver-core/src/main/com/mongodb/ConnectionString.java [hunk 1]
```diff
@@ -475,12 +475,14 @@ private void warnOnUnsupportedOptions(final Map<String, List<String>> optionsMap
     }
 
     private void translateOptions(final Map<String, List<String>> optionsMap) {
+        boolean tlsInsecureSet = false;
```
Comment (jyemin) [inline driver-core/src/main/com/mongodb/ConnectionString.java]: This implementation is really awkward but I couldn't think of a better way without a larger restructuring.

File: driver-core/src/test/unit/com/mongodb/UriOptionsTest.java [hunk 2]
```diff
@@ -42,8 +42,15 @@ public UriOptionsTest(final String filename, final String description, final Str
     public void shouldPassAllOutcomes() {
         assumeFalse(getDefinition().getBoolean("warning", BsonBoolean.FALSE).getValue());
         assumeFalse(getDescription().equals("Arbitrary string readConcernLevel does not cause a warning"));
+        // Skip because Java driver does not support the tlsAllowInvalidCertificates option
+        assumeFalse(getDescription().startsWith("tlsInsecure and tlsAllowInvalidCertificates both present"));
+        assumeFalse(getDescription().startsWith("tlsAllowInvalidCertificates and tlsInsecure both present"));
 
-        testValidOptions();
+        if (getDefinition().getBoolean("valid", BsonBoolean.TRUE).getValue()) {
```
Comment (jyemin) [inline driver-core/src/test/unit/com/mongodb/UriOptionsTest.java]: Now there are test with valid: false so have to handle those

**Comment by jstewart-mongo** [comment]
The new tests LGTM.

File: driver-core/src/main/com/mongodb/ConnectionString.java [hunk 3]
```diff
@@ -498,15 +500,36 @@ private void translateOptions(final Map<String, List<String>> optionsMap) {
...
                 sslInvalidHostnameAllowed = parseBoolean(value, "tlsAllowInvalidHostnames");
+                tlsAllowInvalidHostnamesSet = true;
             } else if (key.equals("sslinvalidhostnameallowed")) {
+                if (tlsInsecureSet) {
+                    throw new IllegalArgumentException("sslInvalidHostnameAllowed set along with tlsInsecure is not allowed");
+                }
                 sslInvalidHostnameAllowed = parseBoolean(value, "sslinvalidhostnameallowed");
+                tlsAllowInvalidHostnamesSet = true;
             } else if (key.equals("tlsinsecure")) {
+                if (tlsAllowInvalidHostnamesSet) {
```
Comment (rozza) [inline driver-core/src/main/com/mongodb/ConnectionString.java]: Could the error checking occur afterwards to DRY this up? eg afterwards check if `tlsInsecureSet` and `tlsAllowInvalidHostnamesSet` at the end and throw?

Comment (jyemin) [reply driver-core/src/main/com/mongodb/ConnectionString.java]: Good idea.  I DRY'd up both this and the ssl/tls check

Comment (rozza) [reply driver-core/src/main/com/mongodb/ConnectionString.java]: Much nicer!

**Pull request closed without merging**